src/middlewared/middlewared/plugins/keychain.py

```diff
@@ -0,0 +1,376 @@
+        return None
+
+    if credential["type"] != "SSH_KEY_PAIR":
+        return None
+
+    if not credential["attributes"]["private_key"]:
+        return None
+
+    return credential
+
+
+def process_ssh_keyscan_output(output):
+    return [" ".join(line.split()[1:]) for line in output.split("\n") if line and not line.startswith("# ")][-1]
+
+
+class KeychainCredentialService(CRUDService):
+
+    class Config:
+        datastore = "system.keychaincredential"
+
+    @accepts(Dict(
+        "keychain_credential_create",
+        Str("name", required=True),
+        Str("type", required=True),
+        Dict("attributes", additional_attrs=True, required=True),
+        register=True,
+    ))
+    async def do_create(self, data):
+        await self._validate("keychain_credential_create", data)
+
+        data["id"] = await self.middleware.call(
+            "datastore.insert",
+            self._config.datastore,
+            data,
+        )
+        return data
+
+    @accepts(
+        Int("id"),
+        Patch(
+            "keychain_credential_create",
+            "keychain_credential_update",
+            ("attr", {"update": True})
+        )
+    )
+    async def do_update(self, id, data):
+        old = await self._get_instance(id)
+
+        new = old.copy()
+        new.update(data)
+
+        await self._validate("keychain_credentials_update", new, id)
+
+        await self.middleware.call(
+            "datastore.update",
+            self._config.datastore,
+            id,
+            new,
+        )
+
+        if new["type"] in ["SSH_KEY_PAIR", "SSH_CREDENTIALS"]:
+            await self.middleware.call("zettarepl.update_tasks")
+
+        return new
+
+    @accepts(Int("id"))
+    async def do_delete(self, id):
+        instance = await self._get_instance(id)
+
+        used_by = await TYPES[instance["type"]].used_by(self.middleware, id)
+        if used_by:
+            raise CallError(f"Unable to delete this credential, it is used by: {', '.join(used_by)}")
+
+        await self.middleware.call(
+            "datastore.delete",
+            self._config.datastore,
+            id,
+        )
+
+    async def _validate(self, schema_name, data, id=None):
+        verrors = ValidationErrors()
+
+        await self._ensure_unique(verrors, schema_name, "name", data["name"], id)
+
+        if data["type"] not in TYPES:
+            verrors.add(f"{schema_name}.type", "Invalid type")
+            raise verrors
+        else:
+            type = TYPES[data["type"]]
+
+            attributes_verrors = validate_attributes(type.credentials_schema, data)
+            verrors.add_child(f"{schema_name}.attributes", attributes_verrors)
+
+        if verrors:
+            raise verrors
+
+        await type.validate_and_pre_save(self.middleware, verrors, f"{schema_name}.attributes", data["attributes"])
+
+        if verrors:
+            raise verrors
+
+    @accepts(Int("id"), Str("type"))
+    async def get_of_type(self, id, type):
```

Comment:
Can we add a docstring like in pool.py for the public methods?

I know the name of the method is most time trivial for us, but having a longer description gives assurance to API users and make the docs look more complete/beautiful :P

Reply:
Sure, I am working on documentation right now.

Reply:
Done (this very method turned out to be private)